### PR TITLE
sdk: notify the room list of a room info change due to late decryption

### PR DESCRIPTION
**Pull request by timokoesters**
This PR fixes https://github.com/element-hq/element-x-ios/issues/1847 for delayed decryptions.
Wrong latest events can still occur when backfill is needed to find the events.

- [ ] Public API changes documented in changelogs (optional)

### CODE REVIEW

File: crates/matrix-sdk/src/sliding_sync/client.rs [hunk 1]
```diff
@@ -53,8 +53,9 @@ impl SlidingSyncResponseProcessor {
         // handle_room_response before this function), so panic is fine.
         assert!(self.response.is_none());
 
-        self.to_device_events =
+        let to_device_events =
             self.client.base_client().process_sliding_sync_e2ee(extensions).await?;
+        self.to_device_events = to_device_events;
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/sliding_sync/client.rs]: nit: you can revert this change please?

File: crates/matrix-sdk/src/sliding_sync/mod.rs [hunk 2]
```diff
@@ -666,7 +666,10 @@ impl SlidingSync {
             }
 
             // Handle the response.
+            // This will batch changes to rooms, so no other service should run at the same time
+            let lock = this.inner.client.locks().sync_service_lock.lock().await;
             let updates = this.handle_response(response, &mut position_guard).await?;
+            drop(lock);
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/sliding_sync/mod.rs:673]: Could the lock be taken only in this block, instead https://github.com/matrix-org/matrix-rust-sdk/blob/main/crates/matrix-sdk/src/sliding_sync/mod.rs#L328-L349? It should be sufficient.

File: crates/matrix-sdk-base/src/client.rs [hunk 3]
```diff
@@ -90,6 +91,7 @@ pub struct BaseClient {
     olm_machine: Arc<RwLock<Option<OlmMachine>>>,
     /// Observable of when a user is ignored/unignored.
     pub(crate) ignore_user_list_changes: SharedObservable<()>,
+    pub(crate) roominfo_update_sender: Arc<StdRwLock<Option<broadcast::Sender<OwnedRoomId>>>>,
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/client.rs:94]: Could it be non-optional so we avoid deferred initialization (aka no more `set_roominfo_update_sender`), and this struct exposes a method to get a receiver, instead?

File: crates/matrix-sdk-base/src/store/mod.rs [hunk 4]
```diff
@@ -217,15 +226,28 @@ impl Store {
 
     /// Lookup the Room for the given RoomId, or create one, if it didn't exist
     /// yet in the store
-    pub fn get_or_create_room(&self, room_id: &RoomId, room_type: RoomState) -> Room {
+    pub fn get_or_create_room(
+        &self,
+        room_id: &RoomId,
+        room_type: RoomState,
+        client: &BaseClient,
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/store/mod.rs:233]: `self` is always ==  `client`, right? Maybe we can avoid it and use `self` instead?

File: crates/matrix-sdk-ui/src/room_list_service/room_list.rs [hunk 5]
```diff
@@ -142,8 +144,41 @@ impl RoomList {
         let stream = stream! {
             loop {
                 let filter_fn = filter_fn_cell.take().await;
-                let (values, stream) = list
-                    .room_list_stream()
+                let (raw_values, mut raw_stream) = list.room_list_stream();
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/room_list_service/room_list.rs:147]: I see you're getting the `room_list_stream()` to maintain `raw_values` (= the list entries) over time. In my opinion it should be just fine to call `list.room_list()` around line 166, so you don't have to maintain the raw list over changes we're receiving (that should simplify the select branch for the raw stream nicely).

File: crates/matrix-sdk-ui/src/room_list_service/room_list.rs [hunk 6]
```diff
@@ -142,8 +144,41 @@ impl RoomList {
...
+                                    for change in &v {
+                                        change.clone().apply(&mut raw_current_values);
+                                    }
+                                    yield v;
+                                } else {
+                                    break;
+                                }
+                            }
+                            room_id = roominfo_update_recv.recv() => {
+                                if let Ok(room_id) = room_id {
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/room_list_service/room_list.rs:165]: style: I think you can rewrite it like that, to spare one level of indent:

```suggestion
                            Ok(room_id) = roominfo_update_recv.recv() => {
```

File: crates/matrix-sdk-ui/src/room_list_service/room_list.rs [hunk 7]
```diff
@@ -142,8 +144,41 @@ impl RoomList {
...
             loop {
                 let filter_fn = filter_fn_cell.take().await;
-                let (values, stream) = list
-                    .room_list_stream()
+                let (raw_values, mut raw_stream) = list.room_list_stream();
+                let mut raw_current_values = raw_values.clone();
+                let client = client.clone();
+                let raw_stream_with_recv = stream! {
+                    loop {
+                        let mut roominfo_update_recv = client.roominfo_update_recv();
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/room_list_service/room_list.rs:152]: Instead of passing the whole `Client` to this function, we could pass it the receiver, and clone it (or not, if not necessary) here.

In general, it's slightly nicer to pass the minimal amount of data as parameters:

- it makes testing easier, as we don't need to mock extra data all over the place,
- it shows precisely what's being used by the called function ("callee")

File: crates/matrix-sdk-ui/src/room_list_service/room_list.rs [hunk 8]
```diff
@@ -142,8 +144,41 @@ impl RoomList {
         let stream = stream! {
             loop {
                 let filter_fn = filter_fn_cell.take().await;
-                let (values, stream) = list
-                    .room_list_stream()
+                let (raw_values, mut raw_stream) = list.room_list_stream();
+                let mut raw_current_values = raw_values.clone();
+                let client = client.clone();
+                let raw_stream_with_recv = stream! {
+                    loop {
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/room_list_service/room_list.rs:151]: That's the second `loop` inside a `loop` :-) Let's extract all this new stream code to another function, please :-)

File: crates/matrix-sdk/src/client/mod.rs [hunk 9]
```diff
@@ -210,6 +210,10 @@ pub(crate) struct ClientLocks {
     /// outside the `OlmMachine`.
     #[cfg(feature = "e2e-encryption")]
     pub(crate) crypto_store_generation: Arc<Mutex<Option<u64>>>,
+
+    /// Ensure that only one service is syncing at a time. Otherwise, they will
+    /// override each other's updates because they batch changes.
+    pub(crate) sync_service_lock: Mutex<()>,
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/client/mod.rs:216]: So we have a `ClientLocks` data structure where to try to put all the locks. Independently of that, the `BaseClient` does already have a `sync_lock`, can we reuse it instead?

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Looks like this is on the right track, congratulations :partying_face: Let's brainstorm together how to test this properly.

File: crates/matrix-sdk-ui/src/room_list_service/room_list.rs (hunk 5 above)
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/room_list_service/room_list.rs:147]: That would be a bug! and a bug doesn't exist until a test case proves it does :sunglasses: :eyes: `</hint>`

File: crates/matrix-sdk-base/src/client.rs (hunk 3 above)
Comment (Hywan) [reply crates/matrix-sdk-base/src/client.rs:94]: You might want to use the client builder to fill this field without an optional value.
Also, we should have the _receiver_ in the client, and the _sender_s in the `RoomInfo`s, right? Except if you use this sender to create the receivers? Anyway, please add a documentation to explain that :-).

File: crates/matrix-sdk-base/src/client.rs [hunk 10]
```diff
@@ -122,9 +124,16 @@ impl BaseClient {
...
+            roominfo_update_sender: Arc::new(StdRwLock::new(None)),
         }
     }
 
+    /// Replaces the sender that triggers updates when room info changes. This
+    /// should only be used once at initialization.
+    pub fn set_roominfo_update_sender(&self, sender: broadcast::Sender<OwnedRoomId>) {
+        *self.roominfo_update_sender.write().unwrap() = Some(sender);
+    }
+
```
Comment (Hywan) [inline crates/matrix-sdk-base/src/client.rs:136]: Should not be needed. Actually, it _must_ be removed, because (i) it's a public method for an internal state, it's bad, (ii) if the sender changes, all receivers are broken, we don't want to change the sender.

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 11]
```diff
@@ -75,6 +76,7 @@ pub struct Room {
     room_id: OwnedRoomId,
     own_user_id: OwnedUserId,
     inner: SharedObservable<RoomInfo>,
+    roominfo_update_sender: Option<broadcast::Sender<OwnedRoomId>>,
```
Comment (Hywan) [inline crates/matrix-sdk-base/src/rooms/normal.rs:79]: You should have a receiver here, not a sender.

**Review by Hywan — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk-ui/src/room_list_service/room_list.rs (hunk 5 above)
Comment (timokoesters) [reply crates/matrix-sdk-ui/src/room_list_service/room_list.rs]: In crates/matrix-sdk-ui/tests/integration/room_list_service.rs I now show that the order is relevant by emitting an update before calling stream.next() and after calling stream.next (Variation 1, Variation 2)

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 12]
```diff
@@ -637,10 +655,15 @@ impl Room {
...
 
-    /// Update the inner summary with the given RoomInfo, and notify
-    /// subscribers.
-    pub fn set_room_info(&self, room_info: RoomInfo) {
+    /// Update the summary with given RoomInfo. This also triggers an update for
+    /// the roominfo_update_recv.
+    pub fn set_room_info(&self, room_info: RoomInfo, trigger_room_list_update: bool) {
         self.inner.set(room_info);
+
+        // Ignore error if no receiver exists
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:663]:
```suggestion
        // Ignore error if no receiver exists.
```

File: crates/matrix-sdk-base/src/sliding_sync.rs [hunk 13]
```diff
@@ -97,7 +97,7 @@ impl BaseClient {
 
         trace!("ready to submit changes to store");
         self.store.save_changes(&changes).await?;
-        self.apply_changes(&changes);
+        self.apply_changes(&changes, true);
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/sliding_sync.rs:100]: It's not correct. If a `SlidingSync` is configured to run both encryption and has configured lists, then a room that has a change for the latest event + a list change because of a room list operation, will receive two updates.

We should only send the update if the sliding sync is not configured to handle room list operations here.

Can you add a test for this, and fix it please? (I think sending a new message after line 733 in the integration test should cause the double update I'm expecting)

File: crates/matrix-sdk/src/client/mod.rs [hunk 14]
```diff
@@ -444,6 +444,13 @@ impl Client {
         self.base_client().session_meta()
     }
 
+    /// Returns a receiver that gets events for each room info update. To watch
+    /// for new events, use `receiver.resubscribe()`. Each event contains the
+    /// room and a boolean whether this event should trigger a room list update.
+    pub fn roominfo_update_receiver(&self) -> &broadcast::Receiver<RoomInfoUpdate> {
+        self.base_client().roominfo_update_receiver()
+    }
+
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/client/mod.rs:453]: I think we could get rid of this method: `base_client()` is public on the `Client`, so it's possible to get the `BaseClient::roominfo_update_receiver` that way, and it's kind of nice that it's semi-hidden by not having it in the public API, since this is a bit of an internal implementation detail. Can you remove it please?

File: crates/matrix-sdk-ui/src/room_list_service/room_list.rs [hunk 15]
```diff
@@ -158,6 +167,51 @@ impl RoomList {
...
+fn merge_stream_and_receiver(
+    mut raw_current_values: Vector<RoomListEntry>,
+    raw_stream: impl Stream<Item = Vec<VectorDiff<RoomListEntry>>>,
+    mut roominfo_update_recv: broadcast::Receiver<RoomInfoUpdate>,
+) -> impl Stream<Item = Vec<VectorDiff<RoomListEntry>>> {
+    let raw_stream_with_recv = stream! {
+        pin_mut!(raw_stream);
+        loop {
+            select! {
+                biased; // Prefer manual updates for easier test code
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/room_list_service/room_list.rs:182]: Let the code breathe a bit :-) Also helps noticing important details, like this biased keyword here.

```suggestion
                biased; // Prefer manual updates for easier test code
                
```

File: crates/matrix-sdk-ui/src/room_list_service/room_list.rs [hunk 16]
```diff
@@ -158,6 +167,51 @@ impl RoomList {
...
+                        if let RoomListEntry::Filled(r) = room {
+                            if r == &update.room_id {
+                                let update = VectorDiff::Set { index, value: raw_current_values[index].clone() };
+                                yield vec![update];
+                                break;
+                            }
+                        }
+                    }
+                }
+                v = raw_stream.next() => {
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/room_list_service/room_list.rs:198]: ditto

```suggestion
                }
                
                v = raw_stream.next() => {
```

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 17]
```diff
@@ -69,13 +70,27 @@ use crate::{
     MinimalStateEvent, OriginalMinimalStateEvent, RoomMemberships,
 };
 
+/// Each time a room's RoomInfo is changed, the roominfo_update_receiver in the
+/// BaseClient will receive this struct containing the update information.
+/// The sender must decide if this update should cause a room list update.
+/// If the sender knows that another action has already caused the room to
+/// update, it should not send another update.
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:77]: This comment is too specific about struct names and field names. Let's make it more generic too, by saying that it describes an update to a `RoomInfo`, with extra data for updates caused by sliding sync that indicates whether an update may be spurious or not.

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 18]
```diff
@@ -69,13 +70,27 @@ use crate::{
...
+#[derive(Debug, Clone)]
+pub struct RoomInfoUpdate {
+    /// The room which was updated.
+    pub room_id: OwnedRoomId,
+    /// Whether this event should trigger the room list to update.
+    pub trigger_room_list_update: bool,
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:83]: I am not a big fan that we're using the use case name here, because the bool's name should describe a fact of the world, not *how* it should be used. It's not a big deal if we don't find a better name, but would be nice to brainstorm one.

File: crates/matrix-sdk-base/src/client.rs [hunk 19]
```diff
@@ -94,6 +94,9 @@ pub struct BaseClient {
     olm_machine: Arc<RwLock<Option<OlmMachine>>>,
     /// Observable of when a user is ignored/unignored.
     pub(crate) ignore_user_list_changes: SharedObservable<()>,
+
+    pub(crate) roominfo_update_sender: broadcast::Sender<RoomInfoUpdate>,
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/client.rs:98]: Can you comment what this is, please?

File: crates/matrix-sdk-base/src/client.rs [hunk 20]
```diff
@@ -94,6 +94,9 @@ pub struct BaseClient {
...
+    pub(crate) roominfo_update_receiver: Arc<broadcast::Receiver<RoomInfoUpdate>>,
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/client.rs:99]: We don't need this field…

File: crates/matrix-sdk-base/src/client.rs [hunk 21]
```diff
@@ -1325,6 +1355,13 @@ impl BaseClient {
             })
             .collect()
     }
+
+    /// Returns a receiver that gets events for each room info update. To watch
+    /// for new events, use `receiver.resubscribe()`. Each event contains the
+    /// room and a boolean whether this event should trigger a room list update.
+    pub fn roominfo_update_receiver(&self) -> &broadcast::Receiver<RoomInfoUpdate> {
+        &self.roominfo_update_receiver
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/client.rs:1363]: …because `self.roominfo_update_sender.subscribe()` will return an owned `Receiver`.

Then we can update the comment to not indicate to use of `resubscribe()`.

File: crates/matrix-sdk-ui/src/room_list_service/room_list.rs [hunk 22]
```diff
@@ -158,6 +167,51 @@ impl RoomList {
...
+/// knows where all rooms are. When the receiver is triggered, a Set operation
+/// for the room position is inserted to the stream.
+fn merge_stream_and_receiver(
+    mut raw_current_values: Vector<RoomListEntry>,
+    raw_stream: impl Stream<Item = Vec<VectorDiff<RoomListEntry>>>,
+    mut roominfo_update_recv: broadcast::Receiver<RoomInfoUpdate>,
+) -> impl Stream<Item = Vec<VectorDiff<RoomListEntry>>> {
+    let raw_stream_with_recv = stream! {
+        pin_mut!(raw_stream);
+        loop {
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/room_list_service/room_list.rs:180]:
```suggestion
        pin_mut!(raw_stream);
        
        loop {
```

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 23]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
 
     Ok(())
 }
+
+static DROP_TODEVICE: StdMutex<bool> = StdMutex::new(true);
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:554]: Can you comment what this does?

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 24]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+fn drop_todevice(response: &mut Bytes) {
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:555]: Let the code breathe :)

```suggestion
static DROP_TODEVICE: StdMutex<bool> = StdMutex::new(true);

fn drop_todevice(response: &mut Bytes) {
```

Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:555]: Please comment what this does. Even if this is test-only code, we need to understand in 6 months what it does at a high-level view.

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 25]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+    let Ok(mut json) = serde_json::from_slice::<serde_json::Value>(response) else {
+        return;
+    };
+    let Some(extensions) = json.get_mut("extensions").and_then(|e| e.as_object_mut()) else {
+        return;
+    };
+    let Some(to_device) = extensions.remove("to_device") else {
+        return;
+    };
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:564]: Any chance we can avoid all the short-circuiting by using `map` and other transformer functions like that? Otherwise, let's comment why it's ok to return in each case.

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 26]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+        info!("Dropping to_device: {to_device}");
+        *response = serde_json::to_vec(&json).unwrap().into();
+        return;
+    }
+}
+
+struct CustomResponder {
+    client: reqwest::Client,
+    response_preprocessor: fn(&mut Bytes),
+}
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:575]: nit: please add newlines after each new struct, impl block, if with a long body, for with a long body, meaningful statement or expression, etc. in this file :-)

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 27]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+        .entries_with_dynamic_adapters(10, alice.roominfo_update_receiver());
+    entries.set_filter(new_filter_all());
+    pin_mut!(stream);
+
+    // Send a message, but the keys won't arrive
+    bob_room.send(RoomMessageEventContent::text_plain("hello world")).await?;
+
+    // Stream only has the initial Reset entry
+    sleep(Duration::from_secs(1)).await;
+    timeout(Duration::from_millis(100), stream.next()).await.unwrap();
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:725]: Instead of two statements, let's increase the timeout duration to 1 second maybe, and remove the sleep.

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 28]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+    assert!(timeout(Duration::from_millis(100), stream.next()).await.is_err());
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:726]: What is this line doing? It seems you'd want to use `assert_pending!` here instead?

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 29]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+    // Latest event is not set yet
+    assert!(matches!(alice_room.latest_event(), None));
+
+    // Now we allow the key to come through
+    *DROP_TODEVICE.lock().unwrap() = false;
+    // Wait for next sync
+    sleep(Duration::from_secs(3)).await;
+
+    // Latest event is set now
+    alice_room.latest_event().unwrap();
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:737]: Can you check its content in an assert too, please?

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 30]
```diff
@@ -198,3 +204,192 @@ async fn test_room_avatar_group_conversation() -> Result<()> {
...
+        .all_rooms()
+        .await
+        .unwrap()
+        .entries_with_dynamic_adapters(10, alice.roominfo_update_receiver());
+    entries.set_filter(new_filter_all());
+    pin_mut!(stream);
+
+    // Stream only has the initial Reset entry
+    timeout(Duration::from_millis(100), stream.next()).await.unwrap();
+    assert!(timeout(Duration::from_millis(100), stream.next()).await.is_err());
```
[the remark this answers is not in the document]
Comment (bnjbvr) [reply testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:726]: How about we only check that the vecdiff is the one that we want, without copying and pasting the entire macro?

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 31]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+    // Now we allow the key to come through
+    *DROP_TODEVICE.lock().unwrap() = false;
+    // Wait for next sync
+    sleep(Duration::from_secs(3)).await;
+
+    // Latest event is set now
+    alice_room.latest_event().unwrap();
+
+    // The stream has a single update
+    timeout(Duration::from_millis(100), stream.next()).await.unwrap();
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:740]: Here too we could check what the update is.

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 32]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+    assert!(timeout(Duration::from_millis(100), stream.next()).await.is_err());
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:741]: and here too we likely want to use `assert_pending!` instead.

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 33]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+        .room_list_service()
+        .all_rooms()
+        .await
+        .unwrap()
+        .entries_with_dynamic_adapters(10, alice.roominfo_update_receiver());
+    entries.set_filter(new_filter_all());
+    pin_mut!(stream);
+    // Initial reset event
+    timeout(Duration::from_millis(100), stream.next()).await.unwrap();
+    assert!(timeout(Duration::from_millis(100), stream.next()).await.is_err());
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:823]: blank lines please :)

Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:823]: also `assert_pending` here?

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 34]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+    assert_eq!(bob_room.state(), RoomState::Joined);
+    // Room update for join
+    timeout(Duration::from_millis(100), stream.next()).await.unwrap();
+    assert!(timeout(Duration::from_millis(100), stream.next()).await.is_err());
+
+    // Send a message, it should arrive
+    bob_room.send(RoomMessageEventContent::text_plain("hello world")).await?;
+
+    sleep(Duration::from_secs(1)).await;
+    alice_room.latest_event().unwrap();
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:842]: Please check the content of the latest event.

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 35]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+    let alice_room = alice.get_room(alice_room.room_id()).unwrap();
+    let bob_room = bob.get_room(alice_room.room_id()).unwrap();
+    bob_room.join().await.unwrap();
+
+    sleep(Duration::from_secs(1)).await;
+    assert_eq!(alice_room.state(), RoomState::Joined);
+    assert!(alice_room.is_encrypted().await.unwrap());
+    assert_eq!(bob_room.state(), RoomState::Joined);
+    // Room update for join
+    timeout(Duration::from_millis(100), stream.next()).await.unwrap();
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:835]: Here and below, please check what the update is + use `assert_pending` when appropriate.

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 36]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+    response_preprocessor: fn(&mut Bytes),
+}
+impl CustomResponder {
+    fn new(response_preprocessor: fn(&mut Bytes)) -> Self {
+        Self { client: reqwest::Client::new(), response_preprocessor }
+    }
+}
+impl wiremock::Respond for CustomResponder {
+    fn respond(&self, request: &wiremock::Request) -> wiremock::ResponseTemplate {
+        let mut req = self.client.request(
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:583]: Please sprinkle a few comments here and there explaining what we're doing in this function (and why we're doing it).

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 37]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+        );
+        for header in &request.headers {
+            for value in header.1 {
+                req = req.header(header.0.to_string(), value.to_string());
+            }
+        }
+        req = req.body(request.body.clone());
+
+        // Run await inside of non-async fn by spawning a new thread and creating a new
+        // runtime. Is there a better way?
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:595]: Could we not implement a `CustomResponder` but instead use a closure that returns an `async` block at line 633?

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 38]
```diff
@@ -538,3 +550,299 @@ async fn test_room_notification_count() -> Result<()> {
...
+    // Send a message, but the keys won't arrive
+    bob_room.send(RoomMessageEventContent::text_plain("hello world")).await?;
+
+    // Stream only has the initial Reset entry
+    sleep(Duration::from_secs(1)).await;
+    timeout(Duration::from_millis(100), stream.next()).await.unwrap();
+    assert!(timeout(Duration::from_millis(100), stream.next()).await.is_err());
+
+    // Latest event is not set yet
+    assert!(matches!(alice_room.latest_event(), None));
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:729]:
```suggestion
    assert!(alice_room.latest_event().is_none());
```

**Review by bnjbvr — CHANGES_REQUESTED** [review]
One small logic bug, and then we need to improve a bit the code comments and documentation. And then we can merge it after the hopefully final round of review. Thanks, it's a bunch of work.

File: crates/matrix-sdk/src/client/mod.rs (hunk 14 above)
Comment (timokoesters) [reply crates/matrix-sdk/src/client/mod.rs]: base_client is only pub(crate) unfortunately

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs (hunk 25 above)
Comment (timokoesters) [reply testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs]: I added some comments, let me know if this needs more changes.

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs (hunk 37 above)
Comment (timokoesters) [reply testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:608]: A closure would also have to return ResponseTemplate and not a future

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs (hunk 27 above)
Comment (timokoesters) [reply testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs]: This breaks the test in some cases, I added some comments

File: crates/matrix-sdk-base/src/sliding_sync.rs (hunk 13 above)
Comment (timokoesters) [reply crates/matrix-sdk-base/src/sliding_sync.rs:100]: If a room has a late decryption that changes the latest_event and at the same time the room is updated by the other code, this will always lead to duplicated updates. It doesn't matter if both updates are made by a single service or two separate services.

I don't think we can fix this without a more involved deduplication algorithm, but this sounds like a rare situation.

Comment (bnjbvr) [reply crates/matrix-sdk-base/src/sliding_sync.rs:100]: I think I see: my proposal doesn't work when there's a latest-event change to the room but no changes in the room list. Well, let's nevermind for now, and get back to it later if it turns out to be an issue.

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 39]
```diff
@@ -538,3 +551,354 @@ async fn test_room_notification_count() -> Result<()> {
...
+                        r = r.append_header(header.0.clone(), header.1.clone());
+                    }
+
+                    let mut bytes = response.bytes().await.unwrap_or_default();
+
+                    // Manipulate the response
+                    response_preprocessor(&mut bytes);
+
+                    r = r.set_body_bytes(bytes);
+                    r
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:632]: This should work:

```suggestion
                    r.set_body_bytes(bytes)
```

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 40]
```diff
@@ -538,3 +551,354 @@ async fn test_room_notification_count() -> Result<()> {
 
     Ok(())
 }
+
+/// Boolean that decides if to_device messages should be dropped.
+static DROP_TODEVICE: StdMutex<bool> = StdMutex::new(true);
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:556]: Instead of global state, could we store this bool in the `CustomResponder` struct?

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 41]
```diff
@@ -538,3 +551,354 @@ async fn test_room_notification_count() -> Result<()> {
...
+    };
+    // Remove to_device field if it exists
+    let Some(to_device) = extensions.remove("to_device") else {
+        return;
+    };
+    if *DROP_TODEVICE.lock().unwrap() {
+        info!("Dropping to_device: {to_device}");
+        *response = serde_json::to_vec(&json).unwrap().into();
+        return;
+    }
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:577]: I think it might be easier to understand by checking the boolean first, and if it's false, return early.

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 42]
```diff
@@ -538,3 +551,354 @@ async fn test_room_notification_count() -> Result<()> {
...
+    let (stream, entries) = alice_sync_service
+        .room_list_service()
+        .all_rooms()
+        .await
+        .unwrap()
+        .entries_with_dynamic_adapters(10, alice.roominfo_update_receiver());
+    entries.set_filter(new_filter_all());
+    pin_mut!(stream);
+
+    // Send a message, but the keys won't arrive
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:737]:
```suggestion
    // Send a message, but the keys won't arrive because to-device events are stripped away from the server's response
```

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs (hunk 31 above)
Comment (bnjbvr) [reply testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:765]: (This comment ^ still valid :P)

File: testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs [hunk 43]
```diff
@@ -538,3 +551,354 @@ async fn test_room_notification_count() -> Result<()> {
...
+
+    sleep(Duration::from_secs(1)).await;
+
+    // Latest event is set now
+    assert_eq!(alice_room.latest_event().unwrap().event_id(), Some(event.event_id));
+
+    // Stream has the room again, but no second event
+    // TODO: Synapse sometimes sends the same event two times. This is the
+    // workaround:
+    assert!(timeout(Duration::from_millis(100), stream.next()).await.unwrap().unwrap().len() > 0);
```
Comment (bnjbvr) [inline testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:890]: Could you use temporary variables here? It's not clear what's the final thing we're checking against...

**Review by bnjbvr — APPROVED** [review]
Yay, let's address the last comments related to the test and merge this. Good job!

**Review by bnjbvr — APPROVED** [review]
Nice, let's get this merged after a rebase :partying_face: Thanks!
